### PR TITLE
New API Modularity - Modules can add Ressources

### LINKED ISSUE #32582: Spike : Modularisation WebAPI
### Description 

Our current implementation does not support modularisation

We wish to know if we need to implement it for the MVP and how 

### Expected attendies

In doc : notion 

- List of risks for module implementation
- Status if is MUST be done for the MVP or later
- List of technical task to do and functional impact ( for example scope def / configuration ). The last one should be done with a PM

### Acceptances criterias 

- Lead Tech & PM validation

### PR DESCRIPTION
**Pull request by 0x346e3730**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Adds the ability for modules to register their own API Platform Resources.
| Type?             | new feature
| Category?         | WS
| BC breaks?        | no
| Deprecations?     | yes
| How to test?      | Install Example API Module and access `/admin-dev/new-api/docs`, see that there is a new `Cat` ressource coming from the module.
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/32582
| Related PRs       | https://github.com/PrestaShop/example-modules/pull/149
| Sponsor company   | PrestaShop

### CODE REVIEW

File: src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php [hunk 1]
```diff
@@ -66,4 +73,24 @@ public function getAlias()
     {
         return 'prestashop';
     }
+
+    public function prepend(ContainerBuilder $container)
+    {
+        foreach ($this->activeModulesPaths as $modulePath) {
+            if (str_ends_with($modulePath, 'api_module')) {
+                $moduleConfigPath = sprintf('%s/config/api_platform', $modulePath);
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php]:
```suggestion
                // Load YAML definition from the config/api_platform folder in the module
                $moduleConfigPath = sprintf('%s/config/api_platform', $modulePath);
```

File: src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php [hunk 2]
```diff
@@ -66,4 +73,24 @@ public function getAlias()
...
+                if (file_exists($moduleConfigPath)) {
+                    $container->prependExtensionConfig('api_platform', ['mapping' => ['paths' => [$moduleConfigPath]]]);
+                }
+                $entitiesRessourcesPath = sprintf('%s/src/Entity', $modulePath);
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php]:
```suggestion
                // Load Doctrine entities that could be used as ApiPlatform DTO resources as well in the src/Entity folder
                $entitiesRessourcesPath = sprintf('%s/src/Entity', $modulePath);
```

File: src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php [hunk 3]
```diff
@@ -66,4 +73,24 @@ public function getAlias()
...
+                if (file_exists($entitiesRessourcesPath)) {
+                    $container->prependExtensionConfig('api_platform', ['mapping' => ['paths' => [$entitiesRessourcesPath]]]);
+                }
+                $moduleRessourcesPath = sprintf('%s/src/ApiPlatform/Resources', $modulePath);
```
Comment (jolelievre) [inline src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php]:
```suggestion
                // Load ApiPLatform DTOs from the src/ApiPlatform/Resources folder
                $moduleRessourcesPath = sprintf('%s/src/ApiPlatform/Resources', $modulePath);
```

**Review by jolelievre — COMMENTED** [review]
Thanks @0x346e3730 

Just a comments about explanation comments, will be easier to understand for future developers

File: src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php [hunk 4]
```diff
@@ -26,16 +26,23 @@
...
 
 /**
  * Adds main PrestaShop core services to the Symfony container.
  */
-class PrestaShopExtension extends Extension
+class PrestaShopExtension extends Extension implements PrependExtensionInterface
 {
+    public function __construct()
+    {
+        $this->activeModulesPaths = (new ModuleRepository(_PS_ROOT_DIR_, _PS_MODULE_DIR_))->getActiveModulesPaths();
```
Comment (FabienPapet) [inline src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php]: I think you can replace their values by using config
That would mean, create some kind of configuration for the bundle: 



```yaml
prestashop: 
    root_dir: !php/const _PS_ROOT_DIR
    module_dir: !php/const _PS_MODULE_DIR_
```

And then you can fetch values inside the container builder

Comment (0x346e3730) [reply src/PrestaShopBundle/DependencyInjection/PrestaShopExtension.php]: It looks like the bundle has no configuration and I'm not sure how to create one for it nor if I should declare the Extension as a service to pass it parameters

**Review by nicosomb — DISMISSED** [review]

**Review by boherm — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by FabienPapet — DISMISSED** [review]
This has legacy constants inside extension, and it is not something we want

**Review by mflasquin — APPROVED** [review]

**Comment by mflasquin** [comment]
tests : https://github.com/mflasquin/ga.tests.ui.pr/actions/runs/5403925506

**Review by FabienPapet — APPROVED** [review]

**Comment by tleon** [comment]
I tested the install of the module and it indeed adds a new endpoint to the api doc. This is all good for me. 
Their is one thing tho. This PR needs to be rebase from develop because it is still using tactician instead of messenger. Once this is done I can give the test a go again before validating for good.

**Comment by tleon** [comment]
I did the rebase on my local setup and everything seems ok. It's a go from me. The rebase needs to be pushed on this branch tho.

**Comment by kpodemski** [comment]
@tleon as a committer you should be allowed to push to @0x346e3730 branch :)

**Comment by tleon** [comment]
Yep I didn't know that but @matks explained it to me. I just did the rebase. So it's all good for me once the CI is green.

**Comment by matks** [comment]
@tleon Dont forget milestone ;)

**Pull request merged**